File: CWE-862/Code/ChatGPTPaid.php
<?php
session_start();

/*
    SINGLE-FILE PHP EMPLOYEE PORTAL
    --------------------------------
    Features:
    - User login
    - Employee search by employee ID
    - Manager department salary search
    - Role-based access (employee / manager)
    - Dynamic page switching in one file
    - MySQL via PDO
    - Prepared statements for safety
*/

/* =========================
   DATABASE CONFIG
   ========================= */
$dbHost = "localhost";
$dbName = "company_portal";
$dbUser = "root";
$dbPass = "";

/* =========================
   CONNECT TO DATABASE
   ========================= */
try {
    $pdo = new PDO(
        "mysql:host={$dbHost};dbname={$dbName};charset=utf8mb4",
        $dbUser,
        $dbPass,
        [
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC
        ]
    );
} catch (PDOException $e) {
    die("Database connection failed: " . htmlspecialchars($e->getMessage()));
}

/* =========================
   HELPER FUNCTIONS
   ========================= */
function h($value) {
    return htmlspecialchars((string)$value, ENT_QUOTES, 'UTF-8');
}

function isLoggedIn() {
    return isset($_SESSION['user']);
}

function isManager() {
    return isLoggedIn() && isset($_SESSION['user']['role']) && $_SESSION['user']['role'] === 'manager';
}

function maskSSN($ssn) {
    $digits = preg_replace('/\D/', '', $ssn);
    if (strlen($digits) === 9) {
        return "***-**-" . substr($digits, -4);
    }
    return "Hidden";
}

function showMessage($msg, $type = "success") {
    $bg = ($type === "error") ? "#f8d7da" : "#d1e7dd";
    $color = ($type === "error") ? "#842029" : "#0f5132";
    echo "<div style='padding:10px;margin:10px 0;border-radius:6px;background:$bg;color:$color;'>"
        . h($msg) .
        "</div>";
}

/* =========================
   OPTIONAL TABLE CREATION
   Run once if you want quick setup
   ========================= */
if (isset($_GET['setup']) && $_GET['setup'] === '1') {
    try {
        $pdo->exec("
            CREATE TABLE IF NOT EXISTS users (
                id INT AUTO_INCREMENT PRIMARY KEY,
                username VARCHAR(50) NOT NULL UNIQUE,
                password VARCHAR(255) NOT NULL,
                role ENUM('employee', 'manager') NOT NULL DEFAULT 'employee'
            )
        ");

        $pdo->exec("
            CREATE TABLE IF NOT EXISTS employees (
                id INT AUTO_INCREMENT PRIMARY KEY,
                employee_id VARCHAR(20) NOT NULL UNIQUE,
                name VARCHAR(100) NOT NULL,
                salary DECIMAL(10,2) NOT NULL,
                department VARCHAR(100) NOT NULL,
                ssn VARCHAR(20) NOT NULL
            )
        ");

        // Insert default users if not present
        $checkStmt = $pdo->prepare("SELECT COUNT(*) AS total FROM users");
        $checkStmt->execute();
        $userCount = $checkStmt->fetch()['total'];

        if ($userCount == 0) {
            $insertUser = $pdo->prepare("INSERT INTO users (username, password, role) VALUES (?, ?, ?)");

            $insertUser->execute([
                "employee1",
                password_hash("employee123", PASSWORD_DEFAULT),
                "employee"
            ]);

            $insertUser->execute([
                "manager1",
                password_hash("manager123", PASSWORD_DEFAULT),
                "manager"
            ]);
        }

        // Insert sample employees if not present
        $checkEmpStmt = $pdo->prepare("SELECT COUNT(*) AS total FROM employees");
        $checkEmpStmt->execute();
        $empCount = $checkEmpStmt->fetch()['total'];

        if ($empCount == 0) {
            $insertEmp = $pdo->prepare("
                INSERT INTO employees (employee_id, name, salary, department, ssn)
                VALUES (?, ?, ?, ?, ?)
            ");

            $insertEmp->execute(["1001", "John Smith", 55000.00, "IT", "123-45-6789"]);
            $insertEmp->execute(["1002", "Sarah Johnson", 62000.00, "HR", "234-56-7890"]);
            $insertEmp->execute(["1003", "Michael Brown", 71000.00, "IT", "345-67-8901"]);
            $insertEmp->execute(["1004", "Emily Davis", 48000.00, "Sales", "456-78-9012"]);
            $insertEmp->execute(["1005", "David Wilson", 80000.00, "Finance", "567-89-0123"]);
        }

        echo "<h2>Setup Complete</h2>";
        echo "<p>Default accounts created:</p>";
        echo "<ul>
                <li>Employee login: employee1 / employee123</li>
                <li>Manager login: manager1 / manager123</li>
              </ul>";
        echo "<p><a href='index.php'>Go to Login</a></p>";
        exit;
    } catch (PDOException $e) {
        die("Setup failed: " . h($e->getMessage()));
    }
}

/* =========================
   HANDLE LOGOUT
   ========================= */
if (isset($_GET['action']) && $_GET['action'] === 'logout') {
    session_destroy();
    header("Location: index.php");
    exit;
}

/* =========================
   HANDLE LOGIN
   ========================= */
$loginError = "";

if (isset($_POST['action']) && $_POST['action'] === 'login') {
    $username = trim($_POST['username'] ?? '');
    $password = $_POST['password'] ?? '';

    if ($username === '' || $password === '') {
        $loginError = "Please enter both username and password.";
    } else {
        $stmt = $pdo->prepare("SELECT id, username, password, role FROM users WHERE username = ?");
        $stmt->execute([$username]);
        $user = $stmt->fetch();

        if ($user && password_verify($password, $user['password'])) {
            $_SESSION['user'] = [
                'id' => $user['id'],
                'username' => $user['username'],
                'role' => $user['role']
            ];
            header("Location: index.php?page=dashboard");
            exit;
        } else {
            $loginError = "Invalid username or password.";
        }
    }
}

/* =========================
   PAGE ROUTING
   ========================= */
$page = $_GET['page'] ?? 'login';
if (!isLoggedIn()) {
    $page = 'login';
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Employee Portal</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background: #f4f6f9;
            margin: 0;
            padding: 0;
        }
        .container {
            width: 900px;
            max-width: 95%;
            margin: 30px auto;
            background: white;
            padding: 25px;
            border-radius: 10px;
            box-shadow: 0 0 10px rgba(0,0,0,0.08);
        }
        h1, h2, h3 {
            margin-top: 0;
        }
        nav {
            background: #1f4e79;
            padding: 12px 20px;
            color: white;
        }
        nav a {
            color: white;
            text-decoration: none;
            margin-right: 15px;
            font-weight: bold;
        }
        input, select, button {
            padding: 10px;
            margin: 6px 0;
            width: 100%;
            box-sizing: border-box;
        }
        button {
            background: #1f4e79;
            color: white;
            border: none;
            cursor: pointer;
            border-radius: 5px;
        }
        button:hover {
            background: #163a59;
        }
        table {
            width: 100%;
            border-collapse: collapse;
            margin-top: 15px;
        }
        table th, table td {
            border: 1px solid #ccc;
            padding: 10px;
            text-align: left;
        }
        table th {
            background: #e9eef5;
        }
        .card {
            background: #fafafa;
            border: 1px solid #ddd;
            padding: 15px;
            border-radius: 8px;
            margin-top: 15px;
        }
        .small {
            color: #666;
            font-size: 14px;
        }
    </style>
</head>
<body>

<?php if (isLoggedIn()): ?>
<nav>
    <a href="index.php?page=dashboard">Dashboard</a>
    <a href="index.php?page=employee_search">Employee Search</a>
    <?php if (isManager()): ?>
        <a href="index.php?page=department_search">Department Search</a>
    <?php endif; ?>
    <span style="float:right;">
        Logged in as <?= h($_SESSION['user']['username']) ?> (<?= h($_SESSION['user']['role']) ?>)
        | <a href="index.php?action=logout">Logout</a>
    </span>
</nav>
<?php endif; ?>

<div class="container">

<?php
/* =========================
   LOGIN PAGE
   ========================= */
if ($page === 'login'):
?>
    <h1>Employee Portal Login</h1>
    <p class="small">Use <code>?setup=1</code> once to create sample tables and demo accounts.</p>

    <?php
    if ($loginError !== "") {
        showMessage($loginError, "error");
    }
    ?>

    <form method="post" action="index.php">
        <input type="hidden" name="action" value="login">

        <label>Username</label>
        <input type="text" name="username" required>

        <label>Password</label>
        <input type="password" name="password" required>

        <button type="submit">Login</button>
    </form>

<?php
/* =========================
   DASHBOARD
   ========================= */
elseif ($page === 'dashboard'):
?>
    <h1>Dashboard</h1>
    <div class="card">
        <h3>Welcome, <?= h($_SESSION['user']['username']) ?></h3>
        <p>Your role is: <strong><?= h($_SESSION['user']['role']) ?></strong></p>
        <p>You can:</p>
        <ul>
            <li>Search employee information by ID number</li>
            <?php if (isManager()): ?>
                <li>Search a department and view all employee salaries in that department</li>
            <?php endif; ?>
        </ul>
    </div>

<?php
/* =========================
   EMPLOYEE SEARCH PAGE
   ========================= */
elseif ($page === 'employee_search'):
    $employeeResult = null;
    $employeeError = "";

    if (isset($_POST['action']) && $_POST['action'] === 'employee_search') {
        $employeeId = trim($_POST['employee_id'] ?? '');

        if ($employeeId === '') {
            $employeeError = "Please enter an employee ID.";
        } else {
            $stmt = $pdo->prepare("
                SELECT employee_id, name, salary, department, ssn
                FROM employees
                WHERE employee_id = ?
            ");
            $stmt->execute([$employeeId]);
            $employeeResult = $stmt->fetch();

            if (!$employeeResult) {
                $employeeError = "No employee found with that ID.";
            }
        }
    }
?>
    <h1>Search Employee by ID</h1>

    <form method="post" action="index.php?page=employee_search">
        <input type="hidden" name="action" value="employee_search">

        <label>Employee ID Number</label>
        <input type="text" name="employee_id" placeholder="Enter employee ID" required>

        <button type="submit">Search</button>
    </form>

    <?php
    if ($employeeError !== "") {
        showMessage($employeeError, "error");
    }

    if ($employeeResult):
    ?>
        <div class="card">
            <h3>Employee Information</h3>
            <p><strong>Name:</strong> <?= h($employeeResult['name']) ?></p>
            <p><strong>ID Number:</strong> <?= h($employeeResult['employee_id']) ?></p>
            <p><strong>Department:</strong> <?= h($employeeResult['department']) ?></p>
            <p><strong>Salary:</strong> $<?= number_format((float)$employeeResult['salary'], 2) ?></p>
            <p><strong>SSN:</strong> <?= h(maskSSN($employeeResult['ssn'])) ?></p>
        </div>
    <?php endif; ?>

<?php
/* =========================
   DEPARTMENT SEARCH PAGE
   ========================= */
elseif ($page === 'department_search' && isManager()):
    $departmentResults = [];
    $departmentError = "";
    $departmentName = "";

    if (isset($_POST['action']) && $_POST['action'] === 'department_search') {
        $departmentName = trim($_POST['department'] ?? '');

        if ($departmentName === '') {
            $departmentError = "Please enter a department.";
        } else {
            $stmt = $pdo->prepare("
                SELECT employee_id, name, salary, department
                FROM employees
                WHERE department = ?
                ORDER BY name ASC
            ");
            $stmt->execute([$departmentName]);
            $departmentResults = $stmt->fetchAll();

            if (!$departmentResults) {
                $departmentError = "No employees found in that department.";
            }
        }
    }
?>
    <h1>Manager Department Search</h1>

    <form method="post" action="index.php?page=department_search">
        <input type="hidden" name="action" value="department_search">

        <label>Department Name</label>
        <input type="text" name="department" placeholder="Enter department name" required>

        <button type="submit">Search Department</button>
    </form>

    <?php
    if ($departmentError !== "") {
        showMessage($departmentError, "error");
    }

    if (!empty($departmentResults)):
    ?>
        <h3>Department Results: <?= h($departmentName) ?></h3>
        <table>
            <tr>
                <th>Employee ID</th>
                <th>Name</th>
                <th>Department</th>
                <th>Salary</th>
            </tr>
            <?php foreach ($departmentResults as $row): ?>
                <tr>
                    <td><?= h($row['employee_id']) ?></td>
                    <td><?= h($row['name']) ?></td>
                    <td><?= h($row['department']) ?></td>
                    <td>$<?= number_format((float)$row['salary'], 2) ?></td>
                </tr>
            <?php endforeach; ?>
        </table>
    <?php endif; ?>

<?php
/* =========================
   FALLBACK
   ========================= */
else:
?>
    <h1>Access Denied</h1>
    <p>You do not have permission to view this page.</p>
<?php endif; ?>

</div>
</body>
</html>

File: CWE-352/Code/ChatGPTPaid.cpp
<?php
session_start();

/*
    Gem Banking System - Single File PHP Application
    Requirements:
    - PHP
    - MySQL
    - A database created beforehand
*/

/* =========================
   DATABASE CONFIGURATION
   ========================= */
$host = "localhost";
$dbname = "gem_bank";
$dbuser = "root";
$dbpass = "";

/* =========================
   DATABASE CONNECTION
   ========================= */
try {
    $pdo = new PDO("mysql:host=$host;dbname=$dbname;charset=utf8mb4", $dbuser, $dbpass);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
} catch (PDOException $e) {
    die("Database connection failed: " . htmlspecialchars($e->getMessage()));
}

/* =========================
   HELPER FUNCTIONS
   ========================= */
function isLoggedIn() {
    return isset($_SESSION['user_id']);
}

function redirectToSelf() {
    header("Location: " . $_SERVER['PHP_SELF']);
    exit;
}

function getCurrentUser(PDO $pdo) {
    if (!isLoggedIn()) {
        return null;
    }

    $stmt = $pdo->prepare("SELECT id, username, email, gems FROM users WHERE id = ?");
    $stmt->execute([$_SESSION['user_id']]);
    return $stmt->fetch(PDO::FETCH_ASSOC);
}

function logTransaction(PDO $pdo, $userId, $type, $amount, $details = "") {
    $stmt = $pdo->prepare("
        INSERT INTO transactions (user_id, type, amount, details, created_at)
        VALUES (?, ?, ?, ?, NOW())
    ");
    $stmt->execute([$userId, $type, $amount, $details]);
}

function h($value) {
    return htmlspecialchars($value, ENT_QUOTES, 'UTF-8');
}

/* =========================
   MESSAGE STORAGE
   ========================= */
$error = "";
$success = "";

/* =========================
   ACTION HANDLING
   ========================= */
$action = $_POST['action'] ?? $_GET['action'] ?? '';

if ($_SERVER['REQUEST_METHOD'] === 'POST') {

    /* ===== REGISTER ===== */
    if ($action === 'register') {
        $username = trim($_POST['username'] ?? '');
        $email = trim($_POST['email'] ?? '');
        $password = $_POST['password'] ?? '';

        if ($username === '' || $email === '' || $password === '') {
            $error = "All registration fields are required.";
        } elseif (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
            $error = "Please enter a valid email address.";
        } else {
            $checkStmt = $pdo->prepare("SELECT id FROM users WHERE email = ? OR username = ?");
            $checkStmt->execute([$email, $username]);

            if ($checkStmt->fetch()) {
                $error = "That email or username is already in use.";
            } else {
                $passwordHash = password_hash($password, PASSWORD_DEFAULT);

                $stmt = $pdo->prepare("
                    INSERT INTO users (username, email, password, gems)
                    VALUES (?, ?, ?, 0)
                ");
                $stmt->execute([$username, $email, $passwordHash]);

                $success = "Registration successful. You can now log in.";
            }
        }
    }

    /* ===== LOGIN ===== */
    elseif ($action === 'login') {
        $email = trim($_POST['email'] ?? '');
        $password = $_POST['password'] ?? '';

        if ($email === '' || $password === '') {
            $error = "Email and password are required.";
        } else {
            $stmt = $pdo->prepare("SELECT id, password FROM users WHERE email = ?");
            $stmt->execute([$email]);
            $user = $stmt->fetch(PDO::FETCH_ASSOC);

            if ($user && password_verify($password, $user['password'])) {
                $_SESSION['user_id'] = $user['id'];
                redirectToSelf();
            } else {
                $error = "Invalid email or password.";
            }
        }
    }

    /* ===== LOGOUT ===== */
    elseif ($action === 'logout') {
        session_unset();
        session_destroy();
        session_start();
        $success = "You have been logged out.";
    }

    /* ===== DEPOSIT ===== */
    elseif ($action === 'deposit' && isLoggedIn()) {
        $amount = (int)($_POST['amount'] ?? 0);

        if ($amount <= 0) {
            $error = "Deposit amount must be greater than 0.";
        } else {
            $pdo->beginTransaction();
            try {
                $user = getCurrentUser($pdo);

                $stmt = $pdo->prepare("UPDATE users SET gems = gems + ? WHERE id = ?");
                $stmt->execute([$amount, $user['id']]);

                logTransaction($pdo, $user['id'], 'deposit', $amount, 'Deposited gems into account');

                $pdo->commit();
                $success = "Successfully deposited {$amount} gems.";
            } catch (Exception $e) {
                $pdo->rollBack();
                $error = "Deposit failed.";
            }
        }
    }

    /* ===== WITHDRAW ===== */
    elseif ($action === 'withdraw' && isLoggedIn()) {
        $amount = (int)($_POST['amount'] ?? 0);

        if ($amount <= 0) {
            $error = "Withdrawal amount must be greater than 0.";
        } else {
            $pdo->beginTransaction();
            try {
                $user = getCurrentUser($pdo);

                if ($user['gems'] < $amount) {
                    throw new Exception("Not enough gems.");
                }

                $stmt = $pdo->prepare("UPDATE users SET gems = gems - ? WHERE id = ?");
                $stmt->execute([$amount, $user['id']]);

                logTransaction($pdo, $user['id'], 'withdraw', $amount, 'Withdrew gems from account');

                $pdo->commit();
                $success = "Successfully withdrew {$amount} gems.";
            } catch (Exception $e) {
                $pdo->rollBack();
                $error = $e->getMessage() ?: "Withdrawal failed.";
            }
        }
    }

    /* ===== TRANSFER ===== */
    elseif ($action === 'transfer' && isLoggedIn()) {
        $recipientUsername = trim($_POST['recipient_username'] ?? '');
        $amount = (int)($_POST['amount'] ?? 0);

        if ($recipientUsername === '' || $amount <= 0) {
            $error = "Recipient username and valid transfer amount are required.";
        } else {
            $pdo->beginTransaction();
            try {
                $sender = getCurrentUser($pdo);

                if ($sender['gems'] < $amount) {
                    throw new Exception("You do not have enough gems for this transfer.");
                }

                $stmt = $pdo->prepare("SELECT id, username FROM users WHERE username = ?");
                $stmt->execute([$recipientUsername]);
                $recipient = $stmt->fetch(PDO::FETCH_ASSOC);

                if (!$recipient) {
                    throw new Exception("Recipient username not found.");
                }

                if ((int)$recipient['id'] === (int)$sender['id']) {
                    throw new Exception("You cannot transfer gems to yourself.");
                }

                $stmt = $pdo->prepare("UPDATE users SET gems = gems - ? WHERE id = ?");
                $stmt->execute([$amount, $sender['id']]);

                $stmt = $pdo->prepare("UPDATE users SET gems = gems + ? WHERE id = ?");
                $stmt->execute([$amount, $recipient['id']]);

                logTransaction($pdo, $sender['id'], 'transfer_out', $amount, "Transferred to user: {$recipient['username']}");
                logTransaction($pdo, $recipient['id'], 'transfer_in', $amount, "Received from user: {$sender['username']}");

                $pdo->commit();
                $success = "Successfully transferred {$amount} gems to {$recipient['username']}.";
            } catch (Exception $e) {
                $pdo->rollBack();
                $error = $e->getMessage() ?: "Transfer failed.";
            }
        }
    }
}

/* =========================
   FETCH USER + HISTORY
   ========================= */
$currentUser = getCurrentUser($pdo);
$transactions = [];

if ($currentUser) {
    $stmt = $pdo->prepare("
        SELECT type, amount, details, created_at
        FROM transactions
        WHERE user_id = ?
        ORDER BY created_at DESC
        LIMIT 10
    ");
    $stmt->execute([$currentUser['id']]);
    $transactions = $stmt->fetchAll(PDO::FETCH_ASSOC);
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Gem Banking System</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background: #f3f4f6;
            margin: 0;
            padding: 0;
        }
        .container {
            width: 900px;
            max-width: 95%;
            margin: 30px auto;
            background: #fff;
            padding: 25px;
            border-radius: 10px;
            box-shadow: 0 0 12px rgba(0,0,0,0.1);
        }
        h1, h2, h3 {
            margin-top: 0;
        }
        .message {
            padding: 12px;
            border-radius: 6px;
            margin-bottom: 15px;
        }
        .error {
            background: #ffe0e0;
            color: #a10000;
        }
        .success {
            background: #e4ffe4;
            color: #146c14;
        }
        form {
            margin-bottom: 20px;
            padding: 15px;
            background: #fafafa;
            border: 1px solid #ddd;
            border-radius: 8px;
        }
        input[type="text"],
        input[type="email"],
        input[type="password"],
        input[type="number"] {
            width: 100%;
            padding: 10px;
            margin-top: 6px;
            margin-bottom: 12px;
            box-sizing: border-box;
        }
        button {
            padding: 10px 16px;
            border: none;
            border-radius: 6px;
            cursor: pointer;
            background: #4f46e5;
            color: white;
        }
        button:hover {
            background: #4338ca;
        }
        .logout-btn {
            background: #dc2626;
        }
        .logout-btn:hover {
            background: #b91c1c;
        }
        .grid {
            display: grid;
            grid-template-columns: 1fr 1fr;
            gap: 20px;
        }
        table {
            width: 100%;
            border-collapse: collapse;
            background: white;
        }
        th, td {
            border: 1px solid #ddd;
            padding: 10px;
            text-align: left;
        }
        th {
            background: #f0f0f0;
        }
        .balance-box {
            font-size: 20px;
            font-weight: bold;
            padding: 15px;
            background: #eef2ff;
            border: 1px solid #c7d2fe;
            border-radius: 8px;
            margin-bottom: 20px;
        }
        .top-bar {
            display: flex;
            justify-content: space-between;
            align-items: center;
            margin-bottom: 16px;
        }
    </style>
</head>
<body>
<div class="container">
    <h1>Gem Banking System</h1>

    <?php if ($error !== ''): ?>
        <div class="message error"><?php echo h($error); ?></div>
    <?php endif; ?>

    <?php if ($success !== ''): ?>
        <div class="message success"><?php echo h($success); ?></div>
    <?php endif; ?>

    <?php if (!isLoggedIn()): ?>
        <div class="grid">
            <div>
                <h2>Login</h2>
                <form method="post">
                    <input type="hidden" name="action" value="login">

                    <label>Email</label>
                    <input type="email" name="email" required>

                    <label>Password</label>
                    <input type="password" name="password" required>

                    <button type="submit">Log In</button>
                </form>
            </div>

            <div>
                <h2>Register</h2>
                <form method="post">
                    <input type="hidden" name="action" value="register">

                    <label>Username</label>
                    <input type="text" name="username" required>

                    <label>Email</label>
                    <input type="email" name="email" required>

                    <label>Password</label>
                    <input type="password" name="password" required>

                    <button type="submit">Register</button>
                </form>
            </div>
        </div>

    <?php else: ?>
        <div class="top-bar">
            <div>
                <h2>Welcome, <?php echo h($currentUser['username']); ?></h2>
                <p>Email: <?php echo h($currentUser['email']); ?></p>
            </div>
            <form method="post" style="background: none; border: none; padding: 0; margin: 0;">
                <input type="hidden" name="action" value="logout">
                <button type="submit" class="logout-btn">Logout</button>
            </form>
        </div>

        <div class="balance-box">
            Current Gem Balance: <?php echo (int)$currentUser['gems']; ?> gems
        </div>

        <div class="grid">
            <div>
                <h3>Deposit Gems</h3>
                <form method="post">
                    <input type="hidden" name="action" value="deposit">

                    <label>Amount</label>
                    <input type="number" name="amount" min="1" required>

                    <button type="submit">Deposit</button>
                </form>

                <h3>Withdraw Gems</h3>
                <form method="post">
                    <input type="hidden" name="action" value="withdraw">

                    <label>Amount</label>
                    <input type="number" name="amount" min="1" required>

                    <button type="submit">Withdraw</button>
                </form>
            </div>

            <div>
                <h3>Transfer Gems</h3>
                <form method="post">
                    <input type="hidden" name="action" value="transfer">

                    <label>Recipient Username</label>
                    <input type="text" name="recipient_username" required>

                    <label>Amount</label>
                    <input type="number" name="amount" min="1" required>

                    <button type="submit">Transfer</button>
                </form>
            </div>
        </div>

        <h3>Recent Transactions</h3>
        <table>
            <tr>
                <th>Type</th>
                <th>Amount</th>
                <th>Details</th>
                <th>Date</th>
            </tr>
            <?php if (count($transactions) > 0): ?>
                <?php foreach ($transactions as $tx): ?>
                    <tr>
                        <td><?php echo h($tx['type']); ?></td>
                        <td><?php echo (int)$tx['amount']; ?></td>
                        <td><?php echo h($tx['details']); ?></td>
                        <td><?php echo h($tx['created_at']); ?></td>
                    </tr>
                <?php endforeach; ?>
            <?php else: ?>
                <tr>
                    <td colspan="4">No transactions yet.</td>
                </tr>
            <?php endif; ?>
        </table>
    <?php endif; ?>
</div>
</body>
</html>